knimin/lib/data_access.py

```diff
@@ -2588,110 +2615,168 @@ def _sample_exists(self, sample_id):
         bool
             Whether the sample exists
         """
-        sql = """SELECT EXISTS (SELECT 1 FROM pm.sample
-                                WHERE sample_id = %s)"""
-        return self._con.execute_fetchone(sql, [sample_id])[0]
+        with TRN:
+            sql = """SELECT EXISTS (SELECT 1 FROM pm.sample
+                                    WHERE sample_id = %s)"""
+            TRN.add(sql, [sample_id])
+            return TRN.execute_fetchlast()
 
-    def _barcode_exists(self, barcode):
-        """Confirms that a barcode exists
+    def _samples_exist(self, sample_ids, exist=True):
+        """Confirms that given sample ID(s) (do not) exist
 
         Parameters
         ----------
-        barcode : str
-            Barcode to check
+        sample_ids : list of str
+            IDs of the sample to check
+        exist : bool, optional, default: True
+            True: checks if exist
+            False: checks if not exist
+
+        Raises
+        ------
+        ValueError
+            If one or more sample IDs do not exist (when exist == True)
+            If one or more sample IDs already exist (when exist == False)
+        """
+        with TRN:
+            sql = """SELECT sample_id
+                     FROM pm.sample
+                     WHERE sample_id IN %s"""
+            TRN.add(sql, [tuple(sample_ids)])
+            res = [x[0] for x in TRN.execute_fetchindex()]
+            if exist:
+                res = set(sample_ids) - set(res)
+                if res:
+                    raise ValueError('Sample ID(s) %s do not exist.' %
+                                     ', '.join(sorted(res)))
+            else:
+                if res:
+                    raise ValueError('Sample ID(s) %s already exist.' %
+                                     ', '.join(sorted(res)))
+
+    def _validate_samples(self, samples):
+        """Confirms that the samples to be created or edited have valid
+        barcodes and study IDs
+
+        Parameters
+        ----------
+        samples : list of dict
+            {barcode : str, optional,
+                Assigns a barcode to the sample
+             study_ids : list of int, optional
+                Associates the sample with one or more studies}
 
         Raises
         ------
         ValueError
-            If the barcode does not exist
+            If at least on study ID does not exist
+            If one or more sample barcodes do not exist
         """
-        sql = """SELECT EXISTS (SELECT 1 FROM barcodes.barcode
-                                WHERE barcode = %s)"""
-        if not self._con.execute_fetchone(sql, [barcode])[0]:
-            raise ValueError('Barcode %s does not exist.' % barcode)
+        with TRN:
+            study_ids = set().union(*[x['study_ids'] for x in samples
+                                      if 'study_ids' in x])
+            for study_id in study_ids:
```

Comment:
I think it would be better to test all the studies at once, as you're doing for the barcodes, to minimize the number of SQL calls and parallelize checks. This way, the user may have a list of 10 studies (that none of them exist), check the first one, fail, then check the next one, fail, etc... up to 10 times. If done like the barcodes test, all the checks happen at once and the user only see 1 error message with all the problems that he has. Does it make sense?